### PR TITLE
[FLINK-2372] Add new FlinkKafkaProducer

### PR DESCRIPTION
**Pull request by rmetzger**
This pull request is reworking Flink's KafkaSink.

I've renamed the class to `FlinkKafkaProducer` to match it with `FlinkKafkaConsumer`.
It uses the new Kafka [Producer API](http://kafka.apache.org/documentation.html#producerapi).
In the documentation, they recommend using the new Producer API:

> As of the 0.8.2 release we encourage all new development to use the new Java producer. This client is production tested and generally both faster and more fully featured than the previous Scala client.

I've also noticed a good performance gain while doing some benchmarks with Flink and Kafka.

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/partitioner/FixedPartitioner.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ */
+package org.apache.flink.streaming.connectors.kafka.partitioner;
+
+/**
+ * A partitioner ensuring that each internal Flink partition ends up in one Kafka partition.
+ *
+ * Note, one Kafka partition can contain multiple Flink partitions.
+ *
+ * Cases:
+ * 	# More Flink partitions than kafka partitions
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/partitioner/FixedPartitioner.java]: These comments should be preformatted `<pre>`

File: flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/partitioner/RichKafkaPartitioner.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+
+/**
+ * Extended Kafka Partitioner.
+ * It contains a prepare() method which is called on each parallel instance.
+ */
+public abstract class RichKafkaPartitioner implements KafkaPartitioner {
+	private static final long serialVersionUID = -4590784174150709918L;
+
+	public abstract void prepare(int parallelInstanceId, int parallelInstances, int[] partitions);
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/partitioner/RichKafkaPartitioner.java]: In all other cases, the method is called `open()`. I think with that we established a bit of a terminology inside Flink, which would be good to follow here.

File: flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/TestFixedPartitioner.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.apache.flink.streaming.connectors.kafka.partitioner.FixedPartitioner;
+
+public class TestFixedPartitioner {
+
+
+	/**
+	 *   		Flink Sinks:		Kafka Partitions
```
Comment (StephanEwen) [inline flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/TestFixedPartitioner.java]: Preformatting

**Comment by StephanEwen** [comment]
Looks pretty good in general. Minor comments inline.

We have now the `Partitioner`, `KafkaPartitioner`, `RichKafkaPartitioner` classes. Can we somehow collapse the later two into one?

**Comment by rmetzger** [comment]
Thank you for the feedback. I will address the concerns.

**Comment by rmetzger** [comment]
I updated the PR and rebased to master (which is adding some commits from the future (at least from GitHubs perspective) )

**Comment by StephanEwen** [comment]
Looks good. Merging this to "master" and "0.10-milestone1"

**Pull request closed without merging**